1st release for AuraWalletCore

khoahm-kypi opened the pull request.

khoahm-kypi committed:
create package core wallet

khoahm-kypi committed:
add dependencies

khoahm-kypi committed:
add enviroment options

khoahm-kypi committed:
create internal storage (  support biomectric ), to stored local data

khoahm-kypi committed:
Added Biomectric Options

khoahm-kypi committed:
create Exception Object, Define error code

khoahm-kypi committed:
Add AuraWalletUtil. This utility class provides methods for retrieving network information, creating internal storage ...

khoahm-kypi committed:
Create Storehouse for Dependency Injection

khoahm-kypi committed:
Add The AuraWallet & AuraTransactionInfo, The AuraWallet represents an abstract interface for managing transactions, balances, and interactions with the Aura Network

khoahm-kypi committed:
Add AuraWalletHelper and Logger, The `AuraWalletHelper` class provides utility methods for various tasks related to the Aura Wallet Core. The `LogInter` class implements a gRPC client interceptor to log gRPC requests and responses.

khoahm-kypi committed:
add AuraWalletImpl, which serves as the concrete implementation of the `AuraWallet` abstract class

khoahm-kypi committed:
add AuraWalletCore, An abstract class representing the core functionality of an Aura wallet.

khoahm-kypi committed:
add the Implementation of AuraWalletCore interface.

khoahm-kypi committed:
Move the classes to correct folder

khoahm-kypi committed:
Add ConfigOption class for customizable core options

khoahm-kypi committed:
Handle config option for GRPC log

khoahm-kypi committed:
remove unused code

khoahm-kypi committed:
Update readme

tuanbass commented on lib/src/cores/utils/aura_wallet_utils.dart:
```diff
@@ -0,0 +1,83 @@
+// Returns the denomination based on the specified environment.
+  static String getDenom(AuraEnvironment environment) {
+    if (environment == AuraEnvironment.euphoria) {
+      return 'ueaura';
```
Do not hard code denom like this. Use a config file (yaml, json..)

khoahm-kypi replied:
Fixed

tuanbass commented on lib/src/helpers/aura_wallet_helper.dart:
```diff
@@ -0,0 +1,136 @@
+    String denom = AuraWalletUtil.getDenom(environment);
+    // Compose the transaction fees
+    final fee = Fee();
+    fee.gasLimit = 200000.toInt64();
```
Why 20000 hard code here??? use a constant class, or  a config file

khoahm-kypi replied:
Fixed

tuanbass commented on lib/src/helpers/aura_wallet_helper.dart:
```diff
@@ -0,0 +1,136 @@
+  /// Returns:
+  ///   - `true` if the private key is valid, `false` otherwise.
+  static bool checkPrivateKey(Uint8List privateKey) {
+    Bip32EccCurve ecc = Bip32EccCurve();
```
Normally you should not create new instance of 3rd party for each check, unless a very lightweight component
By the way, just use 

```dart
return privateKey.leng == 32 && ecc.isPrivate(privateKey)

```
is more expressive

khoahm-kypi replied:
Fixed

tuanbass commented on lib/aura_wallet_core.dart:
```diff
@@ -0,0 +1,63 @@
+  /// Returns an [AuraWallet] instance.
+  Future<AuraWallet> restoreHDWallet({
+    required String passPhrase,
+    String walletName = CONST_DEFAULT_WALLET_NAME,
```
Do you follow the dard standard coding style?[link](https://dart.dev/effective-dart/style#prefer-using-lowercamelcase-for-constant-names)

I'm not sure what the dart community use.
In case we use CAPITAL_CASE, no need prefix CONST

khoahm-kypi replied:
Fixed

tuanbass commented on lib/aura_wallet_core.dart:
```diff
@@ -0,0 +1,63 @@
+  /// Generates a random HD wallet.
+  ///
+  /// Returns an [ComprehensiveWallet] containing the wallet information.
+  Future<ComprehensiveWallet> createRandomHDWallet();
```
- interface: incompatible with the implementation and other methods
where is wallet name??? 
- is it normal to have `create` method and `restore` return different Wallet type?

khoahm-kypi replied:
Fixed

tuanbass commented on lib/src/cores/aura_wallet/aura_wallet_impl.dart:
```diff
@@ -0,0 +1,434 @@
+
+      return response.isSuccessful;
+    } catch (e) {
+      // Handle any exceptions that might occur during the transaction submission.
```
No error message???? How can user (and even dev)  knows what was wrong ???

khoahm-kypi replied:
Fixed

tuanbass commented on lib/src/cores/utils/aura_wallet_utils.dart:
```diff
@@ -0,0 +1,83 @@
+class AuraWalletUtil {
+  // Returns network information based on the specified environment.
+  AuraWalletUtil._();
+  static NetworkInfo getNetworkInfo(AuraEnvironment environment) {
```
use yaml/json to store external configuration like that. Avoid mixing code with hardcoded data

ToanBarcelona1998 replied:
fixed

tuanbass commented on lib/src/cores/aura_wallet/aura_wallet_impl.dart:
```diff
@@ -0,0 +1,434 @@
+    }
+  }
+
+  Future<bool> checkMnemonic({required String mnemonic}) async {
```
duplicate method with util

ToanBarcelona1998 replied:
Fixed

tuanbass commented on lib/src/cores/aura_wallet/aura_wallet_impl.dart:
```diff
@@ -0,0 +1,434 @@
+      String? bech32Address =
+          await Storehouse.storage.getWalletAddress(walletName: walletName);
+
+      List<AuraTransaction>? listSender = await _getListTransactionByAddress(
```
Can we get listSender and listReceiver concurrrently instead of one after one? 
Will decrease waiting time for the user, as the network call is slow

ToanBarcelona1998 replied:
Fixed

tuanbass commented on lib/src/cores/aura_wallet/aura_wallet_impl.dart:
```diff
@@ -0,0 +1,434 @@
+        DateTime? aDate = DateTime.tryParse(a.timestamp);
+        DateTime? bDate = DateTime.tryParse(b.timestamp);
+        if (aDate == null || bDate == null) {
+          return 0; // Skip
```
Should provide a default value, rather return 0(means equal comparison). Potential logic bug for sorting here

ToanBarcelona1998 replied:
Remove and replace by order by parameter.
Fixed.

tuanbass commented on lib/src/cores/aura_wallet/aura_wallet_impl.dart:
```diff
@@ -0,0 +1,434 @@
+          else if (transactionType == AuraTransactionType.recive)
+            "transfer.recipient='$address'"
+        ],
+        pagination: PageRequest(offset: 0.toInt64(), limit: 100.toInt64()),
```
should not have hard code here (limit)

ToanBarcelona1998 replied:
Fixed

tuanbass commented on lib/src/cores/aura_wallet/aura_wallet_impl.dart:
```diff
@@ -0,0 +1,434 @@
+        pagination: PageRequest(offset: 0.toInt64(), limit: 100.toInt64()),
+      );
+
+      final client = auraTx.ServiceClient(networkInfo.gRPCChannel,
```
Seem the network client is created for every cal.
Normally, the network client is big object, and take time to create

ToanBarcelona1998 replied:
Fixed

tuanbass commented on lib/src/cores/aura_wallet/aura_wallet_impl.dart:
```diff
@@ -0,0 +1,434 @@
+
+      return listData;
+    } catch (e) {
+      // Handle any exceptions that might occur while fetching transactions.
```
How can u know what happened when error occurs??

ToanBarcelona1998 replied:
Fixed

tuanbass commented:
General: 
- You need to setup a lint check for enforcement of common best practice

tuanbass commented on lib/src/cores/aura_wallet/aura_wallet_impl.dart:
```diff
@@ -0,0 +1,434 @@
+    }
+
+    // Validate the fee.
+    if (fee != null && fee < 200) {
```
multiple location of hardcoded fee.

ToanBarcelona1998 replied:
Fixed

tuanbass commented on lib/src/cores/aura_wallet/aura_wallet_impl.dart:
```diff
@@ -0,0 +1,434 @@
+  ///
+  /// Returns the response from the smart contract as a string.
+  @override
+  Future<String> makeInteractiveQuerySmartContract({
```
what is interactive query? any different from normal query???
Normally, `interactive` means need intervention from the user

ToanBarcelona1998 replied:
fixed

tuanbass commented on lib/src/cores/aura_wallet/aura_wallet_impl.dart:
```diff
@@ -0,0 +1,434 @@
+  Future<String> makeInteractiveWriteSmartContract({
+    required String contractAddress,
+    required Map<String, dynamic> executeMessage,
+    List<int>? funds,
```
Improvement: Can we provide a way to automatically calc the funds ???

khoahm-kypi replied:
Added "TODO" for handle later

tuanbass commented on lib/src/cores/aura_wallet/aura_wallet_impl.dart:
```diff
@@ -0,0 +1,434 @@
+    if (response.isSuccessful) {
+      return response.txhash;
+    }
+
```
Can other types of exception (such as Network exception) occurs?

ToanBarcelona1998 replied:
Fixed

tuanbass commented on lib/src/cores/aura_wallet/aura_wallet_impl.dart:
```diff
@@ -0,0 +1,434 @@
+  ///
+  /// Throws an [AuraInternalError] with a specific error code and message if any error occurs.
+  @override
+  Future<Tx> makeTransaction({
```
I would name it `transfer` or `send`. There are more type of transactions exist, so this name make audience confuse

ToanBarcelona1998 replied:
fixed

tuanbass commented on lib/src/cores/aura_wallet/aura_wallet.dart:
```diff
@@ -0,0 +1,102 @@
+  const AuraWallet({
+    required this.walletName,
+    required this.bech32Address,
+    required this.environment,
```
In term of api design, as this class doesnt store environment-depend-state, I would prefer to have environment in method call, not the constructor. 
This way, I can reuse the instance for multiple network.

ToanBarcelona1998 replied:
Fixed

tuanbass commented on lib/src/cores/aura_wallet/aura_wallet_impl.dart:
```diff
@@ -0,0 +1,434 @@
+      HttpClient client = HttpClient();
+
+      final request = await client.getUrl(Uri.parse(
+          '$baseUrl/api/v1/transaction?txHash=$txHash&chainid=$chainId'));
```
Should centralize all api path to a class. avoid hardcode/mixing these path with code logic.

ToanBarcelona1998 replied:
Fixed

tuanbass commented on lib/src/cores/aura_wallet/aura_wallet_impl.dart:
```diff
@@ -0,0 +1,434 @@
+    String chainId = AuraWalletUtil.getChainId(environment);
+
+    try {
+      HttpClient client = HttpClient();
```
Should not creat a instance of client for every call, unless you have reason to do it

ToanBarcelona1998 replied:
Fixed

tuanbass commented on lib/src/cores/repo/store_house.dart:
```diff
@@ -0,0 +1,22 @@
+
+/// The `Storehouse` class serves as a central hub for managing dependencies and
+/// providing access to essential configuration elements within the Aura Wallet Core.
+class Storehouse {
```
Seems this class act as a global state for whole application. 
With this implemenation, think about the case that library user need to access different environment/network info. Any conflict????

khoahm-kypi replied:
Fixed

tuanbass commented on lib/src/aura_internal_wallet_ipml.dart:
```diff
@@ -0,0 +1,168 @@
+      }
+
+      // Derive a wallet from the stored passphrase.
+      final Wallet wallet =
```
Should call to `restoreHDWallet` instead of copy/paste code

ToanBarcelona1998 replied:
Because , they don't have the same rules or logic. The "loadStoreWallet" method get wallet from secure storage. It will check user verified two times. The "restoreHDWallet" method must save pass phrase to secure store and it will check user verified one more time. So , I think we should not call to "restoreHDWallet" method

ToanBarcelona1998 replied:
Fixed

tuanbass commented on lib/src/cores/aura_internal_storage.dart:
```diff
@@ -0,0 +1,88 @@
+  }
+
+  Future<String?> readWalletPassPhrase({required String walletName}) async {
+    String? walletAddress = await _storage.read(
```
Should call to getWalletAddress() instead of copy code

ToanBarcelona1998 replied:
Fixed

tuanbass commented on lib/src/cores/aura_internal_storage.dart:
```diff
@@ -0,0 +1,88 @@
+
+  Future<String?> readWalletPassPhrase({required String walletName}) async {
+    String? walletAddress = await _storage.read(
+        key: walletName,
```
Need a transform for `key` instead of use direct wallet name/address as the key 
- Prefixing for different type of value 
- (optional): some kind of encoding to avoid special character

khoahm-kypi replied:
The input parameter walletName is appropriate since the function's purpose is to retrieve the passphrase based on the wallet name. We can consider key transformation for storage security without changing the parameter name

tuanbass commented on lib/src/cores/aura_internal_storage.dart:
```diff
@@ -0,0 +1,88 @@
+    return passPhrase;
+  }
+
+  Future<String?> getWalletAddress({required String walletName}) async {
```
readWalletPassPhrase/getWalletAddress: 
Why the former use "read", the latter use "get"? 

the api should be consistent. `get` is preferable in this case

ToanBarcelona1998 replied:
Fixed

ToanBarcelona1998 committed:
#added environment, refactor comment pr #2

ToanBarcelona1998 committed:
#add error code some cases

ToanBarcelona1998 committed:
#add handle grpc error, some parameter

khoahm-kypi committed:
add config file by dotenv

khoahm-kypi committed:
add constrant values

khoahm-kypi committed:
add The AuraWalletCoreConfigService class, which is responsible for managing configuration options for the Aura Wallet Core SDK. It provides essential configuration elements and initializes the SDK based on the selected environment.

khoahm-kypi committed:
refactor and clean code for aure wallet helper, aura storage, aura transaction info

khoahm-kypi committed:
Add AuraTransactionOrderByType enum and extension,  which provides options for ordering Aura transactions. The enum includes ORDER_BY_UNSPECIFIED, ORDER_BY_ASC, and ORDER_BY_DESC, representing different sorting orders.

khoahm-kypi committed:
refactor: Clean up and optimize AuraWallet/AuraWalletImpl classes

This commit optimizes the AuraWalletImpl class after the code cleanup. The changes include:

- Removed unused imports and dependencies.
- Improved code formatting and consistency.
- Added meaningful comments and documentation to class methods.
- Simplified error handling and added relevant error codes.
- Improved variable naming and readability.

khoahm-kypi committed:
Refactored the AuraWalletCore/AuraWalletCoreImpl classes for improved readability and maintainability.
# - Removed redundant environment assignment by directly using the config service.
# - Updated default wallet name to use a constant.
# - Added better error handling and error code propagation.
# - Simplified the code structure.

khoahm-kypi committed:
Update gitignore

khoahm-kypi committed:
Fix issues "Dont repeat your self" in the code

khoahm-kypi committed:
Update Error Code for delete wallet

ToanBarcelona1998 committed:
Merge branch 'core_wallet' of https://github.com/aura-nw/aura-wallet-core into core_wallet

ToanBarcelona1998 committed:
#remove duplicate code restore wallet

ToanBarcelona1998 committed:
#Fix - Create singleton client one time.

ToanBarcelona1998 committed:
#Remove gaslimit ,minfee options

khoahm-kypi committed:
Refactor Storehouse class for instance-specific configuration and storage

khoahm-kypi committed:
remove ComprehensiveWallet, change to use AuraWallet

khoahm-kypi committed:
Add todo : handle automatically calc the funds
